public/components/event_analytics/explorer/events_views/data_grid.tsx

```diff
@@ -3,148 +3,251 @@
-    if (!loader.current) return;
+  // setSort and setPage are used to change the query and send a direct request to get data
+  const setSort = (sort: EuiDataGridSorting['columns']) => {
+    sortingFields.current = sort;
+    redoQuery(startTime, endTime, rawQuery, timeStampField, sortingFields, pageFields, getEvents);
+  };
 
-    const observer = new IntersectionObserver(
-      (entries) => {
-        if (entries[0].isIntersecting) setLimit((limit) => limit + PAGE_SIZE);
-      },
-      {
-        root: null,
-        rootMargin: '500px',
-        threshold: 0,
-      }
-    );
-    observer.observe(loader.current);
+  const setPage = (page: number[]) => {
+    pageFields.current = page;
+    redoQuery(startTime, endTime, rawQuery, timeStampField, sortingFields, pageFields, getEvents);
+  };
 
-    return () => observer.disconnect();
-  }, [loader]);
+  // creates the header for each column listing what that column is
+  const dataGridColumns = useMemo(() => {
+    if (storedSelectedColumns.length > 0) {
+      const columns: EuiDataGridColumn[] = [];
+      storedSelectedColumns.map(({ name, type }) => {
+        if (name === 'timestamp') {
+          columns.push(DEFAULT_TIMESTAMP_COLUMN);
+        } else if (name === '_source') {
+          columns.push(DEFAULT_SOURCE_COLUMN);
+        } else {
+          columns.push({
+            id: name,
+            display: name,
+            isSortable: true, // TODO: add functionality here based on type
+          });
+        }
+      });
+      return columns;
+    }
+    return [];
+  }, [storedSelectedColumns]);
 
-  const onFlyoutOpen = (docId: string) => {
-    rowRefs.forEach((rowRef) => {
-      rowRef.current?.closeAllFlyouts(docId);
-    });
-  };
+  // used for which columns are visible and their order
+  const dataGridColumnVisibility = useMemo(() => {
+    if (storedSelectedColumns.length > 0) {
+      const columns: string[] = [];
+      storedSelectedColumns.map(({ name }) => {
+        columns.push(name);
+      });
+      return {
+        visibleColumns: columns,
+        setVisibleColumns: (visibleColumns: string[]) => {
+          // TODO: implement with sidebar field order (dragability) changes
+        },
+      };
+    }
+    // default shown fields
+    throw new Error('explorer data grid stored columns empty');
+  }, [storedSelectedColumns]);
+
+  // sets the very first column, which is the button used for the flyout of each row
+  const dataGridLeadingColumns = useMemo(() => {
+    return [
+      {
+        id: 'inspectCollapseColumn',
+        headerCellRender: () => null,
+        rowCellRender: ({ rowIndex }: { rowIndex: number }) => {
+          return (
+            <FlyoutButton
+              ref={null}
+              http={http}
+              key={null}
+              docId={'undefined'}
+              doc={rows[rowIndex % pageFields.current[1]]}
+              selectedCols={explorerFields.queriedFields}
+              timeStampField={timeStampField}
```

Comment:
Should timeStampField be part of the useMemo dependency?

Reply:
timeStampField is just a constant that is passed in as a prop from explorer.tsx. the intention for it being a prop is that if the user wants to define a time field instead of the default one, they can sort by that. but this is not implemented yet. I can still add it to the dependency array, though, for the future